### ISSUE TITLE
Leaking resources/threads in signpipe

### ISSUE THREAD

**Issue by mind04** [PowerDNS/pdns#5322]
This issue is the follow-up of #4992

### Short description
If one of the sign pipe threads is throwing, all the others waiting in poll are stuck forever.
Poll has no time-out here and is relying on the closing of the file descriptor, 

Poll is referring to select() in the notes section.
http://man7.org/linux/man-pages/man2/select.2.html (Multithreaded applications)

~~~A workaround for now is sign-threads=1 in your config.~~~

### Environment
 - Operating system: CentOS 7 x86_64
 - Software version:  rel/auth-4.0.x
 - Software source: self compiled

### Steps to reproduce
1. set sign-thread to a value > 1 (use high values for maximum effect)
2. axfr a domain with a broken record in it ( test.domain.tld A 1.0.0.a for example)

### Expected behaviour
No resource / thread leaking

### Actual behaviour
The number of threads is increasing when pdns is leaking fds and memory.

**Comment by rgacogne** [comment]
I'm probably missing something, but aren't we closing the worker side of a `socketpair` while `poll()` is monitoring the file descriptor corresponding to the other side? The issue mentioned in the manual seems to occur only when the exact same file descriptor than the one monitored by `select()` or `poll()` is closed, right?

**Comment by mind04** [comment]
When i add this small patch in the signpipe destructor (signing-threads=10 and an A record with content 9.9.9.a in the zone to trigger the exception) :
```
--- a/pdns/signingpipe.cc
+++ b/pdns/signingpipe.cc
@@ -95,6 +95,7 @@ ChunkedSigningPipe::ChunkedSigningPipe(const DNSName& signerName, bool mustSign,
 
 ChunkedSigningPipe::~ChunkedSigningPipe()
 {
+  cerr<<"Signpipe destructor start"<<endl;
   delete d_rrsetToSign;
   if(!d_mustSign)
     return;
@@ -107,6 +108,7 @@ ChunkedSigningPipe::~ChunkedSigningPipe()
     pthread_join(tid, &res);
   }
   //cout<<"Did: "<<d_signed<<", records (!= chunks) submitted: "<<d_submitted<<endl;
+  cerr<<"Signpipe destructor end"<<endl;
 }
```
The log of an AXFR looks like this (thread count stable)
```
May 19 11:10:55 AXFR of domain 'wtest.com' initiated by 127.0.0.1
May 19 11:10:55 AXFR of domain 'wtest.com' allowed: client IP 127.0.0.1 is in allow-axfr-ips
May 19 11:10:55 Signing thread died because of std::exception: Parsing record content (try 'pdnsutil check-zone'): unable to parse IP address, strange character: a
May 19 11:11:11 AXFR of domain 'wtest.com' initiated by 127.0.0.1
May 19 11:11:11 AXFR of domain 'wtest.com' allowed: client IP 127.0.0.1 is in allow-axfr-ips
May 19 11:11:11 Signing thread died because of std::exception: Parsing record content (try 'pdnsutil check-zone'): unable to parse IP address, strange character: a
Signpipe destructor start
May 19 11:11:11 Signing thread died because of std::exception: Reading from socket in Signing Pipe loop: Connection reset by peer
May 19 11:11:11 Signing thread died because of std::exception: Reading from socket in Signing Pipe loop: Connection reset by peer
May 19 11:11:11 Signing thread died because of std::exception: Reading from socket in Signing Pipe loop: Connection reset by peer
May 19 11:11:11 Signing thread died because of std::exception: failed in writen2: Broken pipe
May 19 11:11:11 Signing thread died because of std::exception: Reading from socket in Signing Pipe loop: Connection reset by peer
May 19 11:11:11 Signing thread died because of std::exception: Reading from socket in Signing Pipe loop: Connection reset by peer
May 19 11:11:11 Signing thread died because of std::exception: Reading from socket in Signing Pipe loop: Connection reset by peer
May 19 11:11:11 Signing thread died because of std::exception: Reading from socket in Signing Pipe loop: Connection reset by peer
Signpipe destructor end
May 19 11:11:11 TCP Connection Thread died because of STL error: Reading from socket in Signing Pipe loop: Connection reset by peer
```
or sometimes (on average 1 in 10 on my system) like this (thread count +35)
```
May 19 11:11:40 AXFR of domain 'wtest.com' initiated by 127.0.0.1
May 19 11:11:40 AXFR of domain 'wtest.com' allowed: client IP 127.0.0.1 is in allow-axfr-ips
May 19 11:11:40 Signing thread died because of std::exception: Parsing record content (try 'pdnsutil check-zone'): unable to parse IP address, strange character: a
```

**Comment by mind04** [comment]
Fixed in 4.0.5 and 4.1.0
